### DISCUSSION TITLE
Database History using EventSubscriber

### DISCUSSION THREAD

**Discussion by Shivam-Pathak1** [mikro-orm/mikro-orm#2269] [Q&A]
I am fairly new to the concept of LifcycleHooks and EventSubscribers. I am trying to populate a history table that records the changes done on any other table. I have been trying to use EventSubscriber for this:
The idea is that on the onFlush event trigger of my eventSubscriber, I am extracting all the changed values and fields from the changeset and creating an object of the history table and trying to persist it but I see the following error:
`
This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason:
ValidationError: You cannot call em.flush() from inside lifecycle hook handlers
    `

My subscriber looks like this:
```
import { HistoryLog } from 'src/entities/history-log.entity;
export class EntityChangeSubscriber implements EventSubscriber<AnyEntity>  {

  async afterCreate(args: EventArgs<AnyEntity>): Promise<void> {
    console.log('===========1111 AfterCreateHook');
    console.log(args.entity);
  }
  async afterUpdate(args: EventArgs<AnyEntity>): Promise<void> {
    console.log('===========2222 AfterUpdateHook');
    console.log(args.entity);
  }

  async onFlush(args: FlushEventArgs): Promise<void> {
    console.log('===========3333 onFlushHook');
    console.log(args);
    const changeSets = args.uow.getChangeSets();
    const entities = changeSets.map((cs) => cs.entity);
    console.log('Affected entities', entities);
  }

  async afterFlush(args: FlushEventArgs): Promise<void> {
    console.log('===========4444 AfterFlushHook');

    const changeSets = args.uow.getChangeSets();
    console.log('change sets', changeSets);
    changeSets?.map(async (cs) => {
      if (cs.persisted) {
        console.log('print changeset');
        let nextValues = cs.payload ? cs.payload : {};
        let tableName = cs.name;
        let operation = cs.type;
        let rowId = cs.getPrimaryKey()?.toString();
        let modifiedFields = Object.keys(cs.payload);
        let previousValues = cs.originalEntity ? cs.originalEntity : {};
       for (let key in previousValues) {
          if (
            !previousValues.hasOwnProperty(key) ||
            !modifiedFields.includes(key)
          ) {
            delete previousValues[key];
          }
        }
       const historyLog=new HistoryLog({
          operation,
          tableName,
          rowId,
          previousValues,
          nextValues
        });
        console.log(historyLog);
     
        let historyLog1= await args.em.persistAndFlush(historyLog);
        console.log('historyLog1>>> ',historyLog1);
        return historyLog1;
      }
    });
 }
}
```

My mikro-orm-postgres.config.ts registers the subscriber and I can see changeset being loaded
    `subscribers: [new EntityChangeSubscriber()]`

How can I implement this?

**Comment by Shivam-Pathak1** [comment]
Is there a way that I can return the results of EventSubscriber to a service class and do a persist operation there?

**Comment by Shivam-Pathak1** [comment]
I could solve this issue by using a fork before persisting.
I have another issue now. I want to send my userContext object which comes from the network as a Graphql Decorator inside this subscriber and set a few parameters on my HistoryLog object.
I cannot pass UserContext as a parameter to my Entity Manager. I tried setFilterParams which allows me to send objects, but any clue on how can I pass a decorator? Or is there any other approach?

**Reply by jczacharia** [reply]
My use case involves maintaining a History table for my managed `Case`s. I had the same problem when I needed to know which user was executing the Request Context. I tried just having a global variable for the user ID but that led to apparent issues when 2 users make the same request before the first one persists and flushes its data, leading to both historical records showing the second user did the update.  

My solution was to:
1. store the user ID in the request context upon logging in and then 
3. retrieve the user ID when in the Subscription. 

1:
```ts
login(user: User) {
  const context = this.orm.em.getContext() as any;
  context.userId = user.id;
}
```

2:
```ts
@Injectable() // Or just @Subscriber and remove the constructor
export class CaseHistorySubscriber implements EventSubscriber<Case> {
  constructor(private em: EntityManager) {
    this.em.getEventManager().registerSubscriber(this);
  }

  async onFlush(args: FlushEventArgs): Promise<void> {
    const changeSets = args.uow.getChangeSets() as ChangeSet<Case>[];
    for (const cs of changeSets) {
      if (
        (cs.type === ChangeSetType.UPDATE || cs.type === ChangeSetType.CREATE) &&
        cs.entity instanceof Case
      ) {
        const context = args.em.getContext() as any;
        const userId = context.userId as string;
        console.log(userId);
        const record = args.em.create(CaseHistoryEntity, {
          ...cs.entity,
          id: nanoid(),
          user: userId,
          case: cs.entity,
          historyDate: new Date(),
        });
        args.uow.computeChangeSet(record);
      }
    }
  }
}
```
Now with the  `CaseHistory` entity. It's simply all the properties I want to maintain a record of from the `Case` entity.
Such as this:
```ts
@Entity()
export class Case {
  @PrimaryKey()
  id!: string;

  @Property()
  importantData!: string;

  @Property()
  notImportantData!: string;

  @OneToMany(() => CaseHistory, (e) => e.case)
  history = new Collection<CaseHistory>(this);
}
```
```ts
@Entity()
export class CaseHistory {
  @PrimaryKey()
  id!: string;

  @Property()
  importantData!: string;

  @Property({ columnType: 'timestamp with time zone' })
  historyDate!: Date;

  @ManyToOne(() => CaseEntity)
  case!: Case;

  @ManyToOne(() => User)
  user!: User;
}
```
